From 8474cd28c2b5679d808431a4f8d8d9ccd6b87302 Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Sat, 27 Jun 2026 17:32:21 +0400
Subject: [PATCH] fix: gibs: check if killer has quad on vanilla

If the client has the mod and the server is vanilla,
this will give more random speed to gibs.
---
 code/cgame/cg_effects.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/code/cgame/cg_effects.c b/code/cgame/cg_effects.c
index 983d0bdfd..f2a058a7f 100644
--- a/code/cgame/cg_effects.c
+++ b/code/cgame/cg_effects.c
@@ -1119,7 +1119,17 @@ void CG_GibPlayer2( const centity_t *cent, const entityState_t *es,
 	const entityState_t *targEs = &targCent->currentState;
 	const qboolean targEsValid = targCent->currentValid ||
 		targCent == &cg.predictedPlayerEntity;
-	const int killer = es->eventParm;
+
+	const int killerNum = es->eventParm;
+	const centity_t *killerCent = killerNum == cg.snap->ps.clientNum
+		? &cg.predictedPlayerEntity
+		: &cg_entities[ killerNum ];
+	const entityState_t *killerEs = &killerCent->currentState;
+	const qboolean killerEsValid = ( killerCent->currentValid ||
+		killerCent == &cg.predictedPlayerEntity ) &&
+		// With vanilla servers `killer` is always 0
+		// when gibbing a dead body from the body queue.
+		!( targNum >= MAX_CLIENTS && killerNum == 0 );
 
 	vec3_t origin;
 
@@ -1131,6 +1141,11 @@ void CG_GibPlayer2( const centity_t *cent, const entityState_t *es,
 		// Not super necessary but why not.
 		: cgs.g_gibsNewEvGibPlayerProtocol & 0x01
 		? es->eventParm * COMBAT_EV_GIB_PLAYER_ARG_DIVISOR
+		: killerEsValid && killerEs->powerups & ( 1 << PW_QUAD )
+		// Killer has quad: use `MAX_KNOCKBACK`.
+		// TODO fix: also check weapon,
+		// although this will not be always accurate, due to missiles.
+		? 200 * 1000 / COMBAT_PLAYER_MASS
 		// Just use the default knockback speed for 100 damage.
 		: 100 * 1000 / COMBAT_PLAYER_MASS;
 	vec3_t knockbackDir;
@@ -1246,6 +1261,9 @@ void CG_GibPlayer2( const centity_t *cent, const entityState_t *es,
 		CG_Printf(", targ %s%i",
 			targEsValid ? S_COLOR_GREEN : S_COLOR_RED,
 			targNum );
+		CG_Printf(", killer %s%i",
+			killerEsValid ? S_COLOR_GREEN : S_COLOR_RED,
+			killerNum );
 
 		// Yellow means a big difference, but usually it means
 		// that it's an innacuracy that we fixed by using the new protocol,
-- 
2.56.0

